### PR TITLE
enable dump call stack to a buffer

### PR DESCRIPTION
**Pull request by xujuntwt95329**

### CODE REVIEW

File: core/iwasm/aot/aot_runtime.c [hunk 1]
```diff
@@ -3059,14 +3049,96 @@ aot_dump_call_stack(WASMExecEnv *exec_env)
         frame.module_offset = 0;
         frame.func_index = cur_frame->func_index;
         frame.func_offset = 0;
+        frame.func_name_wp =
+            get_func_name_from_index(module_inst, cur_frame->func_index);
 
         if (!bh_vector_append(module_inst->frames.ptr, &frame)) {
             bh_vector_destroy(module_inst->frames.ptr);
```
Comment (wenyongh) [inline core/iwasm/aot/aot_runtime.c:3056]: Should set module_inst->frame.ptr to NULL, or when next time aot_create_call_stack, L3039, the vector will be destroyed again with invalid pointer.

File: core/iwasm/aot/aot_runtime.c [hunk 2]
```diff
@@ -3059,14 +3049,96 @@ aot_dump_call_stack(WASMExecEnv *exec_env)
...
+            if (string_len >= remain_len) {                            \
+                /* Buffer full */                                      \
+                return len;                                            \
+            }                                                          \
+            total_len += string_len;                                   \
+        }                                                              \
+        else {                                                         \
+            total_len += strlen(line_buf);                             \
+        }                                                              \
+    } while (0)
```
Comment (wenyongh) [inline core/iwasm/aot/aot_runtime.c:3084]: The macro is called three times, how about changing it to a function to reduce the footprint? e.g.
```
static uint32
dump_call_stack(const char *line_buf, bool dump_or_print, char *buf, uint32 len)

buf is the place to output the line_buf, not the beginning of total buffer
len is the left size of the buf, not the total size of the buf
return value: the size of buffer printed
```

File: core/iwasm/aot/aot_runtime.c [hunk 3]
```diff
@@ -3059,14 +3049,96 @@ aot_dump_call_stack(WASMExecEnv *exec_env)
...
+
+    snprintf(line_buf, sizeof(line_buf), "\n");
+    PRINT_OR_DUMP();
+
+    while (n < total_frames) {
+        WASMCApiFrame frame = { 0 };
+        uint32 line_length, i;
+
+        if (!bh_vector_get(module_inst->frames.ptr, n, &frame)) {
+            return 0;
```
Comment (wenyongh) [inline core/iwasm/aot/aot_runtime.c:3112]: There are some lines printed already, is it reasonable to return 0?

File: core/iwasm/aot/aot_runtime.h [hunk 4]
```diff
@@ -731,8 +731,24 @@ aot_alloc_frame(WASMExecEnv *exec_env, uint32 func_index);
 void
 aot_free_frame(WASMExecEnv *exec_env);
 
-void
-aot_dump_call_stack(WASMExecEnv *exec_env);
+bool
+aot_create_call_stack(struct WASMExecEnv *exec_env);
+
+/**
+ * @brief dump aot call stack or get the size
```
Comment (wenyongh) [inline core/iwasm/aot/aot_runtime.h:738]: Dump?

File: core/iwasm/aot/aot_runtime.h [hunk 5]
```diff
@@ -731,8 +731,24 @@ aot_alloc_frame(WASMExecEnv *exec_env, uint32 func_index);
...
+ *
+ * @param exec_env the execution environment
+ * @param print print to stdout
```
Comment (wenyongh) [inline core/iwasm/aot/aot_runtime.h:741]: Had better `whether to print to stdout or not`

File: core/iwasm/interpreter/wasm_runtime.c [hunk 6]
```diff
@@ -2560,20 +2560,94 @@ wasm_interp_dump_call_stack(struct WASMExecEnv *exec_env)
             }
         }
 
+        frame.func_name_wp = func_name;
+
+        bh_vector_append(module_inst->frames, &frame);
```
Comment (wenyongh) [inline core/iwasm/interpreter/wasm_runtime.c:2565]: Should check return value, if false, destroy module_inst->frames, and set it to NULL?

File: core/iwasm/interpreter/wasm_runtime.c [hunk 7]
```diff
@@ -2560,20 +2560,94 @@ wasm_interp_dump_call_stack(struct WASMExecEnv *exec_env)
...
+            if (string_len >= remain_len) {                            \
+                /* Buffer full */                                      \
+                return len;                                            \
+            }                                                          \
+            total_len += string_len;                                   \
+        }                                                              \
+        else {                                                         \
+            total_len += strlen(line_buf);                             \
+        }                                                              \
+    } while (0)
```
Comment (wenyongh) [inline core/iwasm/interpreter/wasm_runtime.c:2592]: Same as above, how about make it as function?

File: core/iwasm/interpreter/wasm_runtime.c [hunk 8]
```diff
@@ -2560,20 +2560,94 @@ wasm_interp_dump_call_stack(struct WASMExecEnv *exec_env)
...
+
+    snprintf(line_buf, sizeof(line_buf), "\n");
+    PRINT_OR_DUMP();
+
+    while (n < total_frames) {
+        WASMCApiFrame frame = { 0 };
+        uint32 line_length, i;
+
+        if (!bh_vector_get(module_inst->frames, n, &frame)) {
+            return 0;
```
Comment (wenyongh) [inline core/iwasm/interpreter/wasm_runtime.c:2622]: Same as above, is it reasonable to return 0?

File: core/iwasm/interpreter/wasm_runtime.h [hunk 9]
```diff
@@ -432,8 +432,25 @@ wasm_get_table_inst(const WASMModuleInstance *module_inst, const uint32 tbl_idx)
 }
 
 #if WASM_ENABLE_DUMP_CALL_STACK != 0
-void
-wasm_interp_dump_call_stack(struct WASMExecEnv *exec_env);
+bool
+wasm_interp_create_call_stack(struct WASMExecEnv *exec_env);
+
+/**
+ * @brief dump wasm call stack or get the size
```
Comment (wenyongh) [inline core/iwasm/interpreter/wasm_runtime.h:439]: Dump?

File: core/iwasm/interpreter/wasm_runtime.h [hunk 10]
```diff
@@ -432,8 +432,25 @@ wasm_get_table_inst(const WASMModuleInstance *module_inst, const uint32 tbl_idx)
...
+ *
+ * @param exec_env the execution environment
+ * @param print print to stdout
```
Comment (wenyongh) [inline core/iwasm/interpreter/wasm_runtime.h:442]: Whether to print to stdout or not?

File: core/iwasm/aot/aot_runtime.c (hunk 2 above)
Comment (xujuntwt95329) [reply core/iwasm/aot/aot_runtime.c:3084]: `dump_call_stack` should have the same implementation in both `aot_runtime.c` and `wasm_runtime.c`

how about put it into wasm_runtime_common and rename to `wasm_runtime_dump_call_stack_impl`?

File: core/iwasm/aot/aot_runtime.c (hunk 1 above)
Comment (xujuntwt95329) [reply core/iwasm/aot/aot_runtime.c:3056]: `bh_vector_destroy` doesn't really free the frames.ptr pointer, it just free its internal data and memset the frames.ptr with 0

File: core/iwasm/aot/aot_runtime.c (hunk 2 above)
Comment (wenyongh) [reply core/iwasm/aot/aot_runtime.c:3084]: The function name is a little long, how about `wasm_runtime_dump_line_buf`?

File: core/iwasm/aot/aot_runtime.c (hunk 1 above)
Comment (wenyongh) [reply core/iwasm/aot/aot_runtime.c:3056]: OK, thanks for the notice, had better not change.

File: core/iwasm/aot/aot_runtime.c (hunk 2 above)
Comment (xujuntwt95329) [reply core/iwasm/aot/aot_runtime.c:3084]: Sure, this function will not be exported to developer, we can add some suffix to distinguish, e.g. `wasm_runtime_dump_line_buf_impl`

File: core/iwasm/aot/aot_runtime.c (hunk 3 above)
Comment (xujuntwt95329) [reply core/iwasm/aot/aot_runtime.c:3112]: 0 means error here, even if some lines has been dumped to buffer, the content is not complete, so we return 0 to indicate that the content in buf may be not valid.

File: core/iwasm/interpreter/wasm_runtime.c (hunk 7 above)
Comment (xujuntwt95329) [reply core/iwasm/interpreter/wasm_runtime.c:2584]: Done, thanks~

File: core/iwasm/aot/aot_runtime.h (hunk 4 above)
Comment (xujuntwt95329) [reply core/iwasm/aot/aot_runtime.h]: Done

File: core/iwasm/interpreter/wasm_runtime.c (hunk 6 above)
Comment (xujuntwt95329) [reply core/iwasm/interpreter/wasm_runtime.c]: Done

File: core/iwasm/aot/aot_runtime.c (hunk 3 above)
Comment (wenyongh) [reply core/iwasm/aot/aot_runtime.c:3101]: OK, got it, thanks.

File: core/iwasm/common/wasm_runtime_common.c [hunk 11]
```diff
@@ -4527,22 +4527,87 @@ wasm_externref_retain(uint32 externref_idx)
...
 #endif
 #if WASM_ENABLE_AOT != 0
     if (module_inst->module_type == Wasm_Module_AoT) {
-        aot_dump_call_stack(exec_env);
+        aot_dump_call_stack(exec_env, true, NULL, 0);
     }
 #endif
 }
+
+uint32_t
```
Comment (wenyongh) [inline core/iwasm/common/wasm_runtime_common.c:4571]: Had better change to uint32

File: core/iwasm/common/wasm_runtime_common.c [hunk 12]
```diff
@@ -4527,22 +4527,87 @@ wasm_externref_retain(uint32 externref_idx)
...
+#if WASM_ENABLE_AOT != 0
+    if (module_inst->module_type == Wasm_Module_AoT) {
+        return aot_dump_call_stack(exec_env, false, NULL, 0);
+    }
+#endif
+
+    return 0;
+}
+
+uint32_t
```
Comment (wenyongh) [inline core/iwasm/common/wasm_runtime_common.c:4591]: uint32

File: core/iwasm/common/wasm_runtime_common.c [hunk 13]
```diff
@@ -4527,22 +4527,87 @@ wasm_externref_retain(uint32 externref_idx)
...
+wasm_runtime_dump_call_stack_to_buf(wasm_exec_env_t exec_env, char *buf,
+                                    uint32_t len)
```
Comment (wenyongh) [inline core/iwasm/common/wasm_runtime_common.c:4593]: uint32

Comment (xujuntwt95329) [reply core/iwasm/common/wasm_runtime_common.c]: Done, thanks!

**Pull request merged**